wal.go

```diff
@@ -70,39 +71,129 @@ type WriteAheadLog struct {
 func NewWriteAheadLog(opts *WALOptions) (*WriteAheadLog, error) {
 	walLogFilePrefix := opts.LogDir + "wal"
 
-	firstLogFileName := walLogFilePrefix + ".0.0" // prefix + . {segmentID} + . {starting_offset}
-	file, err := os.OpenFile(firstLogFileName, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	wal := &WriteAheadLog{
+		logFileName:  walLogFilePrefix,
+		maxLogSize:   opts.MaxLogSize,
+		maxSegments:  opts.MaxSegments,
+		log:          opts.Log,
+		syncMaxBytes: opts.SyncMaxBytes,
+		syncTimer:    time.NewTicker(opts.MaxWaitBeforeSync),
+	}
+	err := wal.openExistingOrCreateNew(opts.LogDir)
 	if err != nil {
 		return nil, err
 	}
 
-	fi, err := file.Stat()
+	go wal.keepSyncing()
+
+	return wal, nil
+}
+
+func isDirectoryEmpty(dirPath string) (bool, error) {
+	// Open the directory
+	dir, err := os.Open(dirPath)
 	if err != nil {
-		return nil, err
+		return false, err
 	}
+	defer dir.Close()
 
-	wal := &WriteAheadLog{
-		logFileName: walLogFilePrefix,
-		file:        file,
+	// Read the directory contents
+	fileList, err := dir.Readdir(1) // Read the first entry
+	if err != nil {
+		return false, err
+	}
 
-		maxLogSize: opts.MaxLogSize,
-		logSize:    fi.Size(),
+	// If the list of files is empty, the directory is empty
+	return len(fileList) == 0, nil
```

Comment:
I have added the exception for EOF in the check above. Here, if the error is EOF, the fileList will be empty by default. So this logic should work. Let me know if you find otherwise.